---
title: "Calculatig provenance allele frequencies for GF analyses"
author: "Juliette Archambeau"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    # code_fold: hide
    toc: true
    toc_depth: 4
    toc_float:
       collapsed: false
    number_sections: true
    highlight: textmate
editor_options: 
  chunk_output_type: console
---


<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

<style type="text/css">
div.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 600px;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=F)
options(width = 300)
library(knitr)
library(plyr)
library(dplyr)
library(tidyverse)
library(adespatial)
library(kableExtra)
library(vegan)
```

Like for the calculation of the pairwise FST, **no imputation of the missing data was done here.**

# Using the whole genomic dataset

```{r GenomicDataMunging,eval=F}
# File with the genotype names (clone names)
geno_names <- read.delim2("data/ClonapinBlups523IndPiMASSJuly2019.txt", row.names=1)

# File with the genotype of each clone for each SNP
geno <- read.csv("data/5165snps523genotypesNA.txt", header=FALSE, row.names=1)

# Removing the first two columns with allele info (A,T, G or C)
geno <- geno[,3:dim(geno)[[2]]]

# Give the genotype name for each column of geno
colnames(geno) <- rownames(geno_names)

rm(geno_names)

geno <-  as_tibble(t(geno), rownames = "rownames") %>% 
  mutate(rownames=str_sub(rownames,1,3)) %>% 
  dplyr::rename(prov = rownames) %>% 
  group_by(prov) %>% 
  summarise_all(~sum(., na.rm = TRUE)/((n()-sum(is.na(.)))*2))

  # summarise_all(~sum(., na.rm = TRUE)/(n()*2)) # Previous line that was false because I divided the count of alleles of interest by the total count of allele genotyped in the provenance without taking into account the missing values.

saveRDS(geno,file="data/PreparedDFforGDMandGF/GF/ProvAllelFrequenciesAllSNPs.rds")
```

```{r VisualizeMatrix}
geno <- readRDS(file="data/PreparedDFforGDMandGF/GF/ProvAllelFrequenciesAllSNPs.rds")
geno[1:34,1:20] %>%
  kable(digits=3) %>%  
  kable_styling(font_size = 11,bootstrap_options = c("stripped","hover", "condensed"), full_width = F) %>%
  column_spec(1, bold = T)
```


#  Using the candidate SNPs 

```{r GenomicDataMungingCand,eval=F}
# File with the genotype names (clone names)
geno_names <- read.delim2("data/ClonapinBlups523IndPiMASSJuly2019.txt", row.names=1)

# File with the genotype of each clone for each SNP
geno <- read.csv("data/5165snps523genotypesNA.txt", header=FALSE, row.names=1)

# Removing the first two columns with allele info (A,T, G or C)
geno <- geno[,3:dim(geno)[[2]]]

# Give the genotype name for each column of geno
colnames(geno) <- rownames(geno_names)

rm(geno_names)

# Load candidate SNPs
cand <- readRDS(file="outputs/CandidateSNPs/???")

# Keep only candidate SNPs
geno <- geno[rownames(geno) %in% cand$snp,]

geno <-  as_tibble(t(geno), rownames = "rownames") %>% 
  mutate(rownames=str_sub(rownames,1,3)) %>% 
  dplyr::rename(prov = rownames) %>% 
  group_by(prov) %>% 
  summarise_all(~sum(., na.rm = TRUE)/((n()-sum(is.na(.)))*2))

saveRDS(geno,file="data/PreparedDFforGDMandGF/GF/ProvAllelFrequenciesCandidateSNPs.rds")
```

