---
title: "Formatting, exploring and imputing genomic data"
author: "Juliette Archambeau"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    # code_fold: hide
    toc: true
    toc_depth: 4
    toc_float:
       collapsed: false
    number_sections: true
    highlight: textmate
bibliography: references.bib
editor_options: 
  chunk_output_type: console
---

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

<style type="text/css">
div.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 600px;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 5,fig.height = 4,cache=F)
options(width = 300)
library(knitr)      # CRAN v1.26
library(dplyr)      # CRAN v1.0.0
library(tidyverse)  # CRAN v1.3.0
library(readxl)     # CRAN v1.3.1
library(magrittr)   # CRAN v2.0.1
library(kableExtra) # CRAN v1.1.0
library(genetics)   # to calculate LD
library(xtable)
```

# Load & format all genomic data

```{r LoadGenomicData, warning=F,message=F}
data <- read_excel("data/GenomicData/data_curated_Affx+Inf.xlsx", sheet = 5, col_names = FALSE)
```

```{r DupplicatedPIA3, warning=F,message=F,echo=F,eval=F}

# First genomic file sent by Santi
# --------------------------------

data <- read_excel("data/GenomicData/data_curated_Affx+Inf_PIA3double.xlsx", sheet = 3, col_names = FALSE)

# The clone PIA3 is duplicated.

PIA3 <- data %>%
  filter(...1=="PIA3") %>%
  t() %>%
  set_colnames(c("PIA3_Inf","PIA3_Affx")) %>%
  as_tibble() %>%
  slice(-c(1,2)) %>%
  filter(!PIA3_Affx=="---",!PIA3_Inf=="---") %>%
  mutate(Diff=PIA3_Inf==PIA3_Affx)

# There are 86 differences between the two copies. 

# I initially renamed differently the two.

data <- data %>%  mutate(...1= replace(...1, ...1=="PIA3" & ...2=="only_Inf",  "PIA3a" ),
                 ...1= replace(...1, ...1=="PIA3" & ...2=="only_Affx", "PIA3b" ))

# It is now modified in the new genomic dataset, so this chunk is not useful anymore. 
```


## Dataset with SNP IDs

I create a dataset with the different codes used to refer to the SNP markers: 

  - `original_ID` refers to the original SNP ID.
  
  - `affx_ID` refers to the ID from the Axiom assay (most recent genomic data). 
  
  - `infinium_ID` refers to the ID from the Illumina Infinium assay (older genomic data).

  - `snp_ID` is my own ID that I'm going to create for this paper.
  


```{r CodeDataset}
codes <- data[1:3,] %>% t() %>% as.data.frame() %>% as_tibble()
colnames(codes) <- c("original_ID","affx_ID","infinium_ID")
codes <- codes[-c(1,2),] %>% 
  mutate(snp_ID=paste0("snp_",1:nrow(.)))

# saveRDS(codes,file="data/GenomicData/FormattedData/SNPcodes.rds")

sapply(codes, function(x) sum(is.na(x)))
```

## Genomic data with all data

I create a dataset with all markers (rows) and clones (columns). 

```{r CreatingTotalGenomicDataset1,eval=F}
sub <- data %>% 
  dplyr::select(-1,-2) %>%  # removing the first two columns corresponding to clone ID and assay(s) in which they were genotyped
  dplyr::slice(-c(1,2,3))   # removing the first three rows corresponding to the different SNP IDs

colnames(sub) <- codes$snp_ID # attributing the SNP ID I created to the SNPs
```

I'm going to calculate LD. For that, I download the SNPs I will use latter as reference SNPs (after filtering)

```{r EstimateLD,eval=F}
which.BayPass.selection <- "IS"
snp.sets <- readRDS(file=paste0("outputs/ListSetsSNPs",which.BayPass.selection,".rds"))

# Reference SNPs
sub.LD <- sub %>% 
  dplyr::select(all_of(snp.sets$Ref))

sub.LD[1:10,1:15]  %>% 
  kable() %>%  
  kable_styling(font_size=11,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F)

sub.LD <- makeGenotypes(sub.LD)
LD <- LD(sub.LD)

saveRDS(LD, file="outputs/LD/outputsLDGenetics_Ref.rds")



# Common candidates
sub.LD <- sub %>% 
  dplyr::select(all_of(snp.sets$Com))

sub.LD <- makeGenotypes(sub.LD)
LD <- LD(sub.LD)

saveRDS(LD, file="outputs/LD/outputsLDGenetics_Com.rds")


# Merged candidates
sub.LD <- sub %>% 
  dplyr::select(all_of(snp.sets$Mer))

sub.LD <- makeGenotypes(sub.LD)
LD <- LD(sub.LD)

saveRDS(LD, file="outputs/LD/outputsLDGenetics_Mer.rds")


# Candidates under expected strong selection
sub.LD <- sub %>% 
  dplyr::select(all_of(snp.sets$Mid))

sub.LD <- makeGenotypes(sub.LD)
LD <- LD(sub.LD)

saveRDS(LD, file="outputs/LD/outputsLDGenetics_Mid.rds")
```

```{r LD,eval=F}
snp.sets <- c("Ref","Mer","Mid","Com")
tab <- tibble("SNP sets"=snp.sets, minimum=NA, maximum=NA,mean=NA,median=NA)

# Load LD outputs
for(i in snp.sets){
  LD <- readRDS(file=paste0("outputs/LD/outputsLDGenetics_",i,".rds"))
  tab[tab$`SNP sets`==i,"minimum"] <- range(LD$`R^2`,na.rm=T)[[1]]
  tab[tab$`SNP sets`==i,"maximum"] <- range(LD$`R^2`,na.rm=T)[[2]]
  tab[tab$`SNP sets`==i,"mean"] <- mean(LD$`R^2`,na.rm=T)
  tab[tab$`SNP sets`==i,"median"] <- median(LD$`R^2`,na.rm=T)
}


# Generate the latex table
print(xtable(tab, type = "latex",digits=2), 
      file = "outputs/LD/LDtab.tex", 
      include.rownames=FALSE)
```


```{r CreatingTotalGenomicDataset2,eval=F}
# Converting letters (e.g. A/A, A/G) to numbers (0,1 or 2)
list.alleles <- lapply(colnames(sub), function(x){
  
  vec <- sub[,x]
  
  tab <- table(sub[,x]) %>%  as.matrix() %>%  t() %>% as.data.frame() %>% dplyr::select(-contains("---"))
  
  
  if(length(colnames(tab))==1){ # monomorphic case, e.g. only A/A
    major <- names(which.max(tab))
    
    case_when(vec == major ~ 0) 
    
  } else if(length(colnames(tab))==2){ # case where there is no homozygote with the rare allele, e.g. A/A and A/G
    
    major <- names(which.max(tab))
    mid <- names(which.min(tab))
    
    case_when(vec == major ~ 0,
              vec == mid   ~ 1) 
    
  } else if (length(colnames(tab))==3){ # common case, e.g. A/A, A/G and G/G
    
    if(tab[,1]==tab[,3]){ # to account for cases where there is the same number of A/A and G/G 
      major <- colnames(tab)[1]
      minor <- colnames(tab)[3]
      } else {
    major <- names(which.max(tab[,c(1,3)]))
    minor <- names(which.min(tab[,c(1,3)]))
      }
    mid <- setdiff(names(tab),c(major,minor))  
    
    case_when(vec == major ~ 0,
              vec == minor ~ 2,
              vec == mid   ~ 1) 
  } #else {vec <- rep(3,529)} #  I used this line to check that there was no other case
})

names(list.alleles) <- colnames(sub) # reattribute the SNP ID to list names

df <- list.alleles %>% 
  as_tibble() %>% 
  t()

colnames(df) <- data[4:nrow(data),1] %>%  pull() # attribute clone ID

# saveRDS(df,file="data/GenomicData/FormattedData/TotalGenomicData.rds")
```

```{r, echo=F}
df <- readRDS(file="data/GenomicData/FormattedData/TotalGenomicData.rds")
```


We obtain a dataset of `r ncol(df)` clones (columns) and `r nrow(df)` SNP markers (rows). Here the 10 first rows and 15 first columns:

```{r ShowSubTotalGenomicDataset, echo=F}
df[1:10,1:15]  %>% 
  kable() %>%  
  kable_styling(font_size=11,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
```



# Filtering monomorphic SNPs

```{r IdentifyingMonomorphicSNPs}
mono <- df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(x,na.rm=T)) %>% 
  as.data.frame() %>%
  set_colnames(c("sum")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() %>% 
  filter(sum==0)
```

There are `r nrow(mono)` monomorphic SNPs. I'm going to remove them.


```{r RemovingMonomorphicSNPs}
df <- df[!(row.names(df) %in% mono$snp),]
```

<span style="color: red;">There are `r nrow(df)` SNPs left.</span>


# Identifying duplicates

The first way to identify the duplicated SNPs is by using the function `duplicated`. 

```{r IdentifyPerfectDuplicates}
dup <- df[duplicated(df) | duplicated(df, fromLast=TRUE), ]
```

There are `r nrow(dup)` duplicates (=SNPs that have exactly the same genotype as at least one other SNP).

We can try to determine how many copies there are for each duplicate. 

```{r CountCopiesDuplicates}
count.duplicates <- function(DF){
  x <- do.call('paste', c(DF, sep = '\r'))
  ox <- order(x)
  rl <- rle(x[ox])
  cbind(DF[ox[cumsum(rl$lengths)],,drop=FALSE],count = rl$lengths)
  
}
countdup <- count.duplicates(df %>% as.data.frame()) 
table(countdup$count)
```

This is relevant with what we found before as:

```{r}
2*91 + 3*12 + 4 + 5
```

So it means that there are:

  - <span style="color: red;"> 91 pairs of duplicate SNPs.</span>
  - <span style="color: red;"> 12 groups of 3 duplicate SNPs.</span>
  - <span style="color: red;"> 1 group of 4 duplicate SNPs.</span>
  - <span style="color: red;"> 1 group of 5 duplicate SNPs.</span>

But, there are two situations not accounted for by the function `duplicated`. 

  - First, if two SNPs have exactly the same genotype for all clones, except one clone where one of the two SNPs has a value of "1" for instance, and the other SNP has the value "NA". In this case, the two SNPs will not be considered as duplicates. 
  - Second, if the two SNPs covary perfectly but one have "0" when the other have "2", and vice versa, they will not be counted as duplicates neither. For instance:
      - SNP_1: 0 0 1 1 2 2 
      - SNP_2: 2 2 1 1 0 0 

To add these situations in the count of duplicated SNPs, I selected all the pairs of SNPs with a correlation coefficient of 1. 

```{r IdentifyCorrelatedSNPs,eval=F}
# Invert the genomic matrix
tdf <- df %>% t() %>% as.data.frame() 

# Create a matrix of correlation for all SNPs (5165snps x 5165snps)
matcor <- cor(tdf, use = "pairwise.complete.obs")
matcor[1:10,1:10] %>% 
  kable() %>%  
  kable_styling(bootstrap_options = c("stripped","hover", "condensed"), full_width = F)

# Keep only the upper triangle of the matrix and remove values in the diagnonal
matcor[lower.tri(matcor,diag=T)] <- NA
matcor[1:10,1:10] %>% 
  kable() %>%  
  kable_styling(bootstrap_options = c("stripped","hover", "condensed"), full_width = F)

saveRDS(matcor,file="data/GenomicData/FormattedData/CorrelationMatrixAmongMarkers.rds")

# Counting the number of pairs of SNPs that have a correlation coefficient of 1
length(which(matcor==1))
```


<span style="color: red;">2160 pairs of SNPs have a correlation coefficient of 1.</span>

```{r NamesCorrelatedSNPs,eval=F}
matcor <- readRDS(file="data/GenomicData/FormattedData/CorrelationMatrixAmongMarkers.rds")

# If I want to have the names of the SNPs that have a correlation coefficient of 1
snp.list <- reshape2::melt(matcor) %>% 
  filter(value == 1) %>% 
  dplyr::select(Var1, Var2) %>% 
  dplyr::rename(SNP1=Var1,SNP2=Var2) %>% 
  as_tibble()

# To check that these SNPs have the same genotypes (ignoring NAs) with the following command:
table(df[rownames(df)=="snp_46",],df[rownames(df)=="snp_73",])

# Merge with other SNPs codes
codes_SNP1 <- codes %>% dplyr::rename(SNP1=snp_ID,
                                      original_ID_SNP1=original_ID,
                                      affx_ID_SNP1=affx_ID,
                                      infinium_ID_SNP1=infinium_ID) 
codes_SNP2 <- codes %>% dplyr::rename(SNP2=snp_ID,
                                      original_ID_SNP2=original_ID,
                                      affx_ID_SNP2=affx_ID,
                                      infinium_ID_SNP2=infinium_ID) 

snp.list <- snp.list %>% 
  merge(codes_SNP1,by="SNP1") %>% 
  merge(codes_SNP2,by="SNP2")

# Export the list of SNPs with correlation coefficient of one 
write_csv(snp.list,"data/GenomicData/FormattedData/CorrelatedSnps.csv",append=F)
```


# Removing the ROD population

We remove the clones from the ROD (Rodoiros) populations, as they come from a plantation of material from Central Spain in the Iberian Atlantic region.

```{r RmROD}
df <- df %>% 
  as.data.frame() %>% 
  dplyr::select(-contains("ROD"))
```


# Filtering

## Missing data per clone

```{r MissingDataPerClone, warning=F, fig.height=6,fig.width=10}
tab <- df %>% 
  as_tibble() %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(df)) %>% 
  as.data.frame() %>%
  set_colnames(c("freq.NAs")) %>% 
  rownames_to_column(var = "clone") %>% 
  as_tibble() %>% 
  arrange(freq.NAs) %>% 
  mutate(id=1:nrow(.))

plot(tab$id,tab$freq.NAs,
     pch=20,
     xlab="Clone number",
     ylab="Percentage of missing data per clone")
```

`r nrow(tab[tab$freq.NAs<20,])` clones have the percentage of missing data lower than 20%, `r nrow(tab[tab$freq.NAs<18,])` lower than 18% and `r nrow(tab[tab$freq.NAs<15,])` lower than 15%.

Which threshold do we choose for the missing data per clone?

```{r ThresholdNAperClones}
threshold.NA.per.clone <- 18
```

**<span style="color: red;">We are going to keep the `r nrow(tab[tab$freq.NAs<threshold.NA.per.clone,])` clones that have less than `r threshold.NA.per.clone`% missing data.</span>**

```{r RemoveCloneMissingData}
df <- df %>% 
  dplyr::select(tab[tab$freq.NAs<threshold.NA.per.clone,]$clone)
```

## Minor allele frequencies

```{r MAF, warning=F, fig.height=6,fig.width=10}
tab <- df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(x,na.rm=T)*100/(2*nrow(.))) %>% 
  as.data.frame() %>%
  set_colnames(c("MAF")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() %>% 
  arrange(MAF) %>% 
  mutate(id=1:nrow(.))

plot(tab$id,tab$MAF,
     pch=20,
     xlab="SNP number",
     ylab="Minor allele frequencies")
```

**<span style="color: red;">`r nrow(tab[tab$MAF<1,])` SNPs have MAF < 1%.</span>** 

We remove the SNPs with MAF < 1%.

```{r RmMissingDataMAF}
df <-  df[!(row.names(df) %in% tab[tab$MAF<1,]$snp),]   
```

**<span style="color: red;">There are `r nrow(df)` SNPs left.</span>** 


## Missing data per SNP

```{r MissingDataPerSNP, warning=F, fig.height=6,fig.width=10}
tab <- df %>% 
  t() %>% 
  as_tibble() %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(.)) %>% 
  as.data.frame() %>%
  set_colnames(c("freq.NAs")) %>% 
  rownames_to_column(var = "snp") %>% 
  as_tibble() %>% 
  arrange(freq.NAs) %>% 
  mutate(id=1:nrow(.))

plot(tab$id,tab$freq.NAs,
     pch=20,
     xlab="SNP number",
     ylab="Percentage of missing data per SNP")
```


`r nrow(tab[tab$freq.NAs<20,])` SNPs have the percentage of missing data lower than 20% and `r nrow(tab[tab$freq.NAs<15,])` lower than 15%.

```{r RmMissingDataSNPs}
df <- df[(row.names(df) %in% tab[tab$freq.NAs<20,]$snp),]
```

**<span style="color: red;">We keep SNPs with missing data lower than 20% and so there are `r nrow(df)` SNPs left.</span>**

```{r ReorderColumns}
# We reorder the columns (clone names) in the alphabetical order.
df <- df %>%  dplyr::select(order(colnames(df)))
```


# Export the genomic data

```{r ExportGenomicDataRDS}
nsnp <- nrow(df)
nclone <- ncol(df)

saveRDS(df,
        paste0("data/GenomicData/FormattedData/GenomicData_",nclone,"clones_",nsnp,"snps.rds"))
```


```{r ExportGenomicDataCSV}
df.csv <- codes %>% 
  merge(df %>% rownames_to_column(var="snp_ID"),by="snp_ID") %>% 
  dplyr::rename(juliette_snp_ID=snp_ID)

write_csv(df.csv,
        paste0("data/GenomicData/FormattedData/GenomicData_",nclone,"clones_",nsnp,"snps.csv"),
        append=FALSE)
```

# Information useful for the paper

## How many SNPs genotyped in each assay?

We want to know how many SNPs were genotyped in the Infiunium assay, in the Affymetric assay and in both. 

```{r NumberSNPgenotypedAssays}
nb <- sapply(df.csv[,1:4], function(x) sum(!(is.na(x))))

df.noNAS <- df.csv[,c(1,3,4)] %>% drop_na()
```

Among the `r nb[[1]]` SNPs used in this paper (after filtering), **`r nb[[3]]` were genotyped in the Affymetrix assay**, **`r nb[[4]]` in the Infinium assay** and **`r nrow(df.noNAS)` were genotyped by both assays**.

## How many clones genotyped in each assay?

We want to know the same for the clones, i.e. how many clones were genotypes by each assay and by both.

```{r NumberClonesGenotypedAssays}
tab.clon <- data[4:nrow(data),1:2] %>% 
  setNames(c("clon","assay")) %>% 
  filter(clon %in% colnames(df))

# sapply(tab.clon, function(x) sum(is.na(x))) # (to check) No NAs, great .

table(tab.clon$assay)
```

All clones were genotyped with both the Affymetrix and Infinium assays.

## Average and max number of missing data per clone

Frequency of missing data per clone:

```{r MissingDataPerclone}
tab <- df %>% 
  sapply(function(x) sum(is.na(x))*100/nrow(df)) %>% 
  as.data.frame() %>%
  set_colnames(c("freq.NAs")) %>% 
  rownames_to_column(var = "clone") %>% 
  as_tibble() %>% 
  arrange(desc(freq.NAs)) %>% 
  mutate(id=1:nrow(.))

# Visualization
plot(tab$id,tab$freq.NAs,
     pch=20,
     xlab="Clone number",
     ylab="Percentage of missing data per clone")

# Table ordered in decreasing order
# so clones with the highest frequencies of missing data are first
tab[c(1:10),1:2] %>%
  kable() %>%  
  kable_styling(font_size=11,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
  
```

The **maximum frequency of missing data per clone** is **`r round(max(tab$freq.NAs),2)`%** and the **average frequency of missing data per clone** is **`r round(mean(tab$freq.NAs),2)`%**.

```{r GenerateTableLatexSuppInfoNbClonesPerProv, echo=F, eval=F}
# We generate a table for the Supplementary Information of the paper, with the number of clones (i.e. trees) per provenance:
tibble(Provenance=str_sub(colnames(df),1,3),clon=colnames(df)) %>% 
  group_by(Provenance) %>% 
  count(name="Nb.clones") %>% 
  kable("latex", booktabs=T) 
```

# Imputation of missing data

Percentage of missing data in the entire genomic dataset: `r round(sum(is.na(df)) * 100 / (dim(df)[1]*dim(df)[2]),2)`%.

We can use two ways to impute the missing values:

  1. Method from the [RDA vignette](https://popgen.nescent.org/2018-03-27_RDA_GEA.html) but that we are not going to use here. We can impute the missing values using the most common allele at each SNP across all individuals.

  2. Method that we are going to use: Imputation of the missing values using the most common allele at each SNP within the gene pool. So here, the genotypes are classified according to the main gene pool they belong to (although we aknowledge that some genotypes are admixed and belong to different gene pools).



## First method (not used)

Using the most common allele at each SNP across all individuals.

```{r 1stMethodImputation, eval=F}
gen.imp <- apply(df, 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))
sum(is.na(gen.imp)) # No NAs
```


## Second method (the one we use in the paper)

Using the most common genotype at each SNP within the main gene pool this genotype belong to.

We load the dataset in which there is the proportion of belonging of each genotype to a given gene pool [from @jaramillo2015molecular].

```{r LoadingPropAncestry}
prop <- readRDS(file="data/PhenoEnvPopStructureDataset.RDS") %>% 
  dplyr::select(clon,prov,paste0(rep("Q",6),1:6),max.Q) %>%  
  unique() %>% 
  drop_na()
```

We check whether some clones in the genomic dataset we want to use for the following analyses have no population structure data.

```{r IdentifyClonesNoPopStructure}
setdiff(colnames(df),prop$clon)
```

This is ok, we know the proportion of membership to each gene pool for each clone we want to use in the following analyses! 

<!-- Comment: If we use the threshold of 20% missing data per clone, we would have seen that one selected clone (OLO17) has no population structure data. However, if we look at the other clones of the OLO population, all mainly belong to the French Atlantic gene pool (`Q4`): -->

```{r OLOclones,echo=F}
prop.OLO <- prop %>% dplyr::filter(prov == "OLO")

# prop.OLO %>% 
#   kable() %>%
#   kable_styling(bootstrap_options = c("hover", "condensed"), full_width = F) %>%
#   column_spec(1, bold = T)
```

<!-- Among the `r nrow(prop.OLO)` clones of the OLO population, the proportion of belonging from the French Atlantic gene pool ranges from `r min(prop.OLO$Q4)` to `r max(prop.OLO$Q4)`. So there is a high probability that OLO17 also belong mainly from the French Atlantic gene pool. That's why we consider in the following analyses that OLO17 mainly belong from the French Atlantic gene pool. -->

```{r MergePopStructureWithGenomicData}
mainGP <- prop %>% dplyr::select(clon,max.Q) # extract the main gene pool for each clone.

# if threshold missing data per clone < 20%, we add the main  gene pool of OLO17
if(threshold.NA.per.clone>18) mainGP[nrow(mainGP) + 1,] <- c("OLO17","Q4") 

gen.imp <- df %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("clon") %>%  
  left_join(mainGP,by="clon") %>% 
  dplyr::select(clon,max.Q,everything())

gen.imp[1:10,1:10] %>%
  kable() %>%
  kable_styling(bootstrap_options = c("hover", "condensed"), full_width = F) %>%
  column_spec(1, bold = T)
```

```{r CheckingImputation1, echo=F, eval=F}
# If we want to check that the imputation worked fine
  # Find some SNPs with NAs
subset <- gen.imp[gen.imp$max.Q=="Q5",]
sample(colnames(subset)[colSums(is.na(subset)) > 0],3,replace=F) #extract some SNPs with NAs
sum(is.na(subset)) # 14,965 NAs in Q5 gene pool

table(subset[,"snp_12238"]) # NAs of this SNP should take the value '2' after imputation.
table(subset[,"snp_10182"]) # NAs of this SNP should take the value '1' after imputation.
table(subset[,"snp_7628"])  # NAs of this SNP should take the value '2' after imputation.
```


We impute the missing data:

```{r NAsImputation}
# Replace NAs by (1) subseting by GP, (2) imputing NAs, (3) reassembling the GP.
for(i in unique(gen.imp$max.Q)){
  subset <- gen.imp[gen.imp$max.Q==i,]
  subset <- apply(subset[,3:ncol(subset)], 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))
  gen.imp[gen.imp$max.Q==i,3:ncol(gen.imp)] <- subset
}
```


```{r CheckingImputation2, echo=F, eval=F}
# If we want to check that the imputation worked fine
  # Look at the SNPs that had NAs before imputation
subset <- gen.imp[gen.imp$max.Q=="Q5",]
sum(is.na(subset)) # 0 NAs now!

table(subset[,"snp_12238"]) # great! (NAs of this SNP has taken the values '2' after imputation)
table(subset[,"snp_10182"]) # great! (The missing value of this SNP has taken the value '1' after imputation)
table(subset[,"snp_7628"])  # great! (The missing value of this SNP has taken the value '2' after imputation)
```

Formating the dataset (reattributing clone names, clones in columns and snp in rows, etc.):

```{r ReformatImputedDataset}
rownames(gen.imp) <- gen.imp$clon
gen.imp <- gen.imp %>% 
  dplyr::select(-clon,-max.Q) %>% 
  t() %>% 
  as.data.frame()

# Visualization:
gen.imp[1:10,1:10] %>%
  kable() %>%
  kable_styling(bootstrap_options = c("hover", "condensed"), full_width = F) %>%
  column_spec(1, bold = T)
```

 
## Exporting the dataset

We save this dataset to use it in the univariate genotype-environment association analysis with <span style="font-variant:small-caps;">BayPass</span>:

```{r ExportGenomicDataImputedRDS}
nsnp <- nrow(gen.imp)
nclone <- ncol(gen.imp)

saveRDS(gen.imp,
        paste0("data/GenomicData/FormattedData/ImputedGenomicData_",nclone,"clones_",nsnp,"snps.rds"))
```


```{r ExportGenomicDataImputedCSV}
gen.imp.csv <- codes %>% 
  merge(gen.imp %>% rownames_to_column(var="snp_ID"),by="snp_ID") %>% 
  dplyr::rename(juliette_snp_ID=snp_ID)

write_csv(gen.imp.csv,
        paste0("data/GenomicData/FormattedData/ImputedGenomicData_",nclone,"clones_",nsnp,"snps.csv"),
        append=FALSE)
```


# References