---
title: "Formatting data for GDM and GF analyses"
author: "Juliette Archambeau"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    # code_fold: hide
    toc: true
    toc_depth: 4
    toc_float:
       collapsed: false
    number_sections: true
    highlight: textmate
editor_options: 
  chunk_output_type: console
---


<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

<style type="text/css">
div.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 600px;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=F)
options(width = 300)
library(knitr)
library(plyr)
library(dplyr)
library(hierfstat)
library(tidyverse)
library(adespatial)
library(kableExtra)
library(vegan)
```


# Load data

We used the genomic dataset in which there is missing data (i.e. no imputation of the missing data).


```{r LoadGenomicData}
df <- readRDS(file="data/GenomicData/FormattedData/GenomicData_454clones_9817snps.rds")
```

```{r LoadSNPsSets}
snp.sets <- readRDS(file="outputs/ListSetsSNPs.rds")
```



# Provenance allele frequencies for GF analyses

```{r CalculateAlleleFrequenciesForGF, eval=F}
list.freq <- sapply(snp.sets, function(x){ 

  df %>% 
  filter(rownames(.) %in% x) %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var="prov") %>% 
  mutate(prov=str_sub(prov,1,3)) %>% 
  group_by(prov) %>% 
  summarise_all(~sum(., na.rm = TRUE)/((n()-sum(is.na(.)))*2))
  
  # summarise_all(~sum(., na.rm = TRUE)/(n()*2)) 
  # This line that was false because I divided the count of alleles of interest
  # by the total count of allele genotyped in the provenance without taking into account the missing values.

  }, USE.NAMES = TRUE,simplify=FALSE)


saveRDS(list.freq,file="outputs/GF/ListAlleleFrequencies.rds")
```


```{r VisualizeSubsetMatrix, eval=F}
list.freq[[1]][1:34,1:20] %>%
  kable(digits=3) %>%  
  kable_styling(font_size = 11,bootstrap_options = c("stripped","hover", "condensed"), full_width = F) %>%
  column_spec(1, bold = T)
```

# Pairwise $F_{ST}$ for GDM


```{r LoadNeededFunctions}
source("scripts/Functions/PairwiseWCfstSmallSamp.R")
```

```{r FormatGenomicData}
df[df ==1] <- 12
df[df ==2] <- 22
df[df ==0] <- 11

df <- df %>% 
  t() %>% 
  as.data.frame() %>% 
  dplyr::mutate(prov=substr(row.names(.), 0, 3)) %>% 
  dplyr::select(prov, everything())

df[1:10,1:10] %>%  
  kable() %>%  
  kable_styling(font_size = 11,bootstrap_options = c("stripped","hover", "condensed"), full_width = F) %>%
  column_spec(1, bold = T)
```

To estimate the pairwise $F_{ST}$, we use the **Weir and Cockerham method**, with the function `pairwise.WCfst` of the `hierfstat` package. We could have used the function `gene.dist` (with the option `WC84`), as in [Gougherty et al. (2020)](https://www.biorxiv.org/content/10.1101/2020.02.28.961060v1.abstract). `gene.dist` keeps only the lower triangle of the $F_{ST}$ matrix, while `pairwise.WCfst` keeps the whole matrix.


```{r CompatureFstMatrixRefSNPs,echo=F,eval=T}
# To save time, we are going to calculate now the pairwise Fst matrix of reference SNPs

fst <- df %>% pairwise.WCfst(diploid=TRUE)

saveRDS(list.fst,file="outputs/GDM/FstMatrices/PairwiseFstReferenceSNPs.rds")
```


```{r ComputePairwiseFstMatrices, eval=F}
list.fst <- sapply(snp.sets, function(x){ 
  
  df %>% 
  dplyr::select(prov,all_of(x)) %>%
  pairwise.WCfst(diploid=TRUE)
    
}, USE.NAMES = TRUE,simplify=FALSE)

saveRDS(list.fst,file="outputs/GDM/FstMatrices/ListPairwiseFst.rds")
```

```{r LoadPairwiseFstMatrices}
list.fst <- readRDS(file="outputs/GDM/FstMatrices/ListPairwiseFst.rds")
```


## Negative values

```{r NegValuestoZeroFstMatrix}
NegValues <- sapply(list.fst, function(x){
  length(x[which(x <0)])
}, USE.NAMES = TRUE,simplify=FALSE)
```

There are some negative values in the pairwise $F_{ST}$ matrices:

  - `r NegValues[["Ref"]]` in the matrix of the reference SNPs.
  - `r NegValues[["Mer"]]` in the matrix of the merged candidates.
  - `r NegValues[["Mid"]]` in the matrix of the candidates under expected strong selection.
  - `r NegValues[["Com"]]` in the matrix of the common candidates.
  
  
It generally means there is more variation within than between provenances and is well known to result from uneven sample sizes.

See here some comments on negative values obtained when calculating pairwise Fst:

  - https://www.biostars.org/p/77372/
  
  - https://www.researchgate.net/post/How-the-Fst-value-is-with-negative-value
  
In [Gougherty et al. 2020](https://www.biorxiv.org/content/10.1101/2020.02.28.961060v1.abstract) : "Any pairwise $F_{ST}$ values less than zero were assigned a value of zero."

I asked Matt Fitzpatrick the following question: "I saw also that in Gougherty et al. (2020), you did not scale the Fst values but you assigned the value of 0 to any pairwise Fst values less than zero. Is there a reason to prefer one method over the other?".

Here Matt Fitzpatrick's answer: "The only reason we have done this is simply to facilitate model convergence, as you mention. I will double check with Andy G as I would be surprised if he didn't have to rescale his Fst values 0-1."


## Scaling the $F_{ST}$ matrices

To scale the $F_{ST}$ matrices between 0 and 1 to help convergence, Matt Fitzpatrick said (in his email 05/02/2021) that he substracted the minimum value and then divided by the maximum value in Fitzpatrick & Keller (2015), such as:

$$x_{new} = \frac{x- x_{min}}{x_{max}} $$

In @ingvarsson2020genome: "Genetic distances between populations, from the pairwise FST matrix, were scaled to lie between 0 and 1 by **subtracting the minimum value and then dividing by the maximum value** (Fitzpatrick & Keller, 2015). Scaling was performed to enable comparisons between reference and associated SNPs that displayed different ranges of observed FST values."

```{r ScalingFSTmatrixLikeFitzpatrick}
sapply(list.fst, function(x){
  
  mat <- (x-min(x,na.rm=T))/max(x,na.rm=T)
  range(mat,na.rm=T)
  
}, USE.NAMES = TRUE,simplify=FALSE)

```

In our case, this method is not working because some $F_{ST}$ values are still higher than 1.
 
So we use another method in the paper: we standardize the $F_{ST}$ values in the following way:

$$x_{new} = \frac{x- x_{min}}{x_{max}-x_{min}} $$
This was proposed here: https://www.statisticshowto.com/probability-and-statistics/normal-distributions/normalized-data-normalization/


```{r ScalingFSTmatrixStandardization}
ScaledMatrices <- sapply(list.fst, function(x){
  
 (x-min(x,na.rm=T))/(max(x,na.rm=T)-min(x,na.rm=T))
  
}, USE.NAMES = TRUE,simplify=FALSE)


sapply(ScaledMatrices, function(x){range(x,na.rm=T)}, USE.NAMES = TRUE,simplify=FALSE)

saveRDS(ScaledMatrices,file="outputs/GDM/FstMatrices/ListPairwiseFstScaled.rds")
```

## Visualization

```{r visualizeMatrixFunctions}
source("scripts/Functions/corpmat.R")
col <- colorRampPalette(c("#BB4444", "#EE9988", "#FFFFFF", "#77AADD", "#4477AA"))


VizMatrix <- function(x) {

  p.mat <- corpmat(x)

  corrplot::corrplot(x, method="color", col=col(200),
                   type="upper", order="hclust", 
                   addCoef.col = "black", # Add coefficient of correlation
                   tl.col="black", tl.srt=23, #Text label color and rotation
                   # Combine with significance
                   p.mat = p.mat, sig.level = 0.05, insig = "blank", number.cex =0.8,tl.cex = 0.8,
                   # hide correlation coefficient on the principal diagonal
                   diag=FALSE)
}
```

### Reference SNPs

```{r VizPairwiseFstMatrixRef, fig.height=14, fig.width=14}
VizMatrix(ScaledMatrices[["Ref"]])
```

### Merged candidate SNPs

```{r VizPairwiseFstMatrixMer, fig.height=14, fig.width=14}
VizMatrix(ScaledMatrices[["Mer"]])
```

### Candidate SNPs under expected strong selection 

```{r VizPairwiseFstMatrixMid, fig.height=14, fig.width=14}
VizMatrix(ScaledMatrices[["Mid"]])
```

### Common candidate SNPs

```{r VizPairwiseFstMatrixCom, fig.height=14, fig.width=14}
VizMatrix(ScaledMatrices[["Com"]])
```


`r knitr::opts_chunk$set(eval = F)`
