---
title: 'Univariate genetic-environment association: <span style="font-variant:small-caps;">BayPass</span>'
author: "Juliette Archambeau"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    # code_fold: hide
    toc: true
    toc_depth: 4
    toc_float:
       collapsed: false
    number_sections: true
    highlight: textmate
bibliography: references.bib 
editor_options: 
  chunk_output_type: console
---

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

<style type="text/css">
div.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 600px;
}
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 5,fig.height = 4,cache=FALSE)
options(width = 300)
library(knitr)
library(vegan)  # to investigate correlations among predictors
library(psych)  # to plot the pairs panels
library(tidyverse)
library(corrplot)
library(kableExtra)
```

```{r BayPassUtils}
source("~/Bureau/baypass_2.2/utils/baypass_utils.R")
```

In this document, we test on our data the different models proposed by <span style="font-variant:small-caps;">BayPass</span>. The results appeared to be the most robust with the important sampling models. Even if nothing was said in the <span style="font-variant:small-caps;">BayPass</span> manual on whether we should run IS one covariate at a time or not, we decided to do so. So, the scripts that were used for the paper are at the end of this document (section: `IS with one covariate at a time`).

# Download <span style="font-variant:small-caps;">BayPass</span>

I downloaded [here](http://www1.montpellier.inra.fr/CBGP/software/baypass/download.html) the version 2.2 of <span style="font-variant:small-caps;">BayPass</span>.

I extracted the `tar.gz` file in `~/Bureau` with:

```{bash ExtractBayPasstargz, eval=F}
cd ~/Bureau/
tar -zxvf baypass_2.2.tar.gz
```

To compile <span style="font-variant:small-caps;">BayPass</span>, I used the gfortran free compiler (the command should automatically produce an executable called `g_baypass`).

```{bash CompileBaypass, eval=F}
cd baypass_2.2/sources/
make clean all FC=gfortran
```

To execute <span style="font-variant:small-caps;">BayPass</span>, e.g. here the `help` command:

```{bash}
cd ~/Bureau/baypass_2.2/sources
./g_baypass -help
```


# Load and format data
 
## Genomic data

We load the genomic dataset: 

```{r LoadGenomicData}
geno <- readRDS(file="data/GenomicData/FormattedData/ImputedGenomicData_457clones_9833snps.rds") %>% t() %>% as.data.frame()
geno[1:10,1:10]
```

There are `r dim(geno)[2]` SNPs and `r dim(geno)[1]` clones (i.e. genotypes). Missing data were imputed based on the main gene pool of the clone (using the most common allele at each SNP).

The genomic data for <span style="font-variant:small-caps;">BayPass</span> needs to be allele counts for each SNP in each of the provenance. 

```{r CalculateAlleleCounts, eval=F}
count.2nd.allele = function(x) {2*length(x)-sum(x)}
stats = c('sum', 'count.2nd.allele')

df <- geno %>% mutate(prov=str_sub(row.names(geno),1,3)) %>% 
  select(prov,everything()) %>% 
  group_by(prov) %>% 
  dplyr::summarize_all(.funs=stats)  # very long

saveRDS(df,file=paste0("outputs/CandidateSNPs/baypass/PreFile_",dim(geno)[2],"snps",dim(geno)[1],"clones_ImpNABayPass.rds"))
```

```{r ArrangeAlleleCounts}
df <- readRDS(file=paste0("outputs/CandidateSNPs/baypass/PreFile_",dim(geno)[2],"snps",dim(geno)[1],"clones_ImpNABayPass.rds"))

df[1:10,1:10] %>% 
  kable() %>%  
  kable_styling(font_size=12,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F)

dfsum1  <- df %>% column_to_rownames(var="prov") %>% select(contains("sum")) %>%  t()   # count of the first allele
dfsum2  <- df %>% column_to_rownames(var="prov") %>% select(contains("count")) %>%  t() # count of the second allele

# dfsum1 <- df %>%  select(contains("sum")) %>%  t()  # count of the first allele
# dfsum2 <- df %>%  select(contains("count")) %>% t() # count of the second allele
# colnames(dfsum2) <- colnames(dfsum1) <- df$prov

colnames(dfsum1) <- paste0(df$prov,"1")
colnames(dfsum2) <- paste0(df$prov,"2")

rownames(dfsum1) <- str_sub(rownames(dfsum1),1,-5)
rownames(dfsum2) <- str_sub(rownames(dfsum2),1,-18)

identical(rownames(dfsum1), rownames(dfsum2)) 

DF <- cbind(dfsum1,dfsum2) %>% as_tibble %>% 
  select(sort(tidyselect::peek_vars()))

DF[1:10,1:10] %>% 
  kable() %>%  
  kable_styling(font_size=12,
                bootstrap_options = c("stripped","hover", "condensed"), full_width = F)
```

```{r SaveDF, eval=F}
write.table(DF, file = paste0("outputs/CandidateSNPs/baypass/",dim(geno)[2],"snps",dim(geno)[1],"clones_ImpNABayPass.rds"), sep = " ",
            row.names = FALSE, col.names = FALSE) # works either with .txt or no extension
```

## Environmental data

Here the set of environmental covariates we use in this paper (see report `reports/EnvironmentalData/BuilingEnvPhenoPopStructureDataset.Rmd`):

Burned area: `BurnedArea`.

Topographic variable:`TRI`: topographic ruggedness index

Soil variables:

  - `water_top` = sand content in the top layers of the soil
  - `depth_roots` = Depth available to roots

Climatic variables representing episodic events:

  - `bio5`: max temperature of the warmest month (°C)
  - `bio6`: min temperature of the coldest month (°C)
  - `bio12`: annual precipitation (mm)
  - `bio15`: precipitation Seasonality (Coefficient of Variation)

```{r SelectedEnvVariables}
# Names  of the environmental variables of interest in the dataset:
select.var <- c("bio5WC_prov",
                "bio6WC_prov",
                "bio12WC_prov",
                "bio15WC_prov",
                "water_top_prov",
                "depth_roots_prov",
                "TRI",
                "BurnedArea")
```


```{r LoadEnvPopStructureData}
env <- readRDS(file="data/PhenoEnvPopStructureDataset.RDS") %>% 
  dplyr::select(prov,all_of(select.var)) %>% 
  dplyr::rename(bio5="bio5WC_prov",
                bio6="bio6WC_prov",
                bio12="bio12WC_prov",
                bio15="bio15WC_prov",
                water_top="water_top_prov",
                depth_roots="depth_roots_prov") %>% 
  unique()

identical(env$prov, df$prov) # provenances are in the same order as in the genomic data

# Names of the variables we are going to use in the document
select.var <- c("bio5",
                "bio6",
                "bio12",
                "bio15",
                "water_top",
                "depth_roots",
                "TRI",
                "BurnedArea")

env <- env %>% dplyr::select(-prov) %>%  t() %>% round(3)

# One table per covariate
for(i in select.var){
  write.table(env[i,], file = paste0("outputs/CandidateSNPs/baypass/EnvironmentalVariables/",i), sep = " ",
            row.names = FALSE, col.names = FALSE)
}
```
